Userland/Services/SpiceAgent/SpiceAgent.h

```diff
@@ -1,129 +1,74 @@
+    // Indicates where the message has come from
+    enum class Port : u32 {
         Client = 1,
-        Server
-    };
-
-    struct [[gnu::packed]] ChunkHeader {
-        u32 port {};
-        u32 size {};
-    };
-
-    struct [[gnu::packed]] Message {
-        u32 protocol;
-        u32 type;
-        u64 opaque;
-        u32 size;
-        u8 data[];
-    };
-
-    enum class MessageType {
-        MouseState = 1,       // server -> client
-        MonitorsConfig,       // client -> agent|server
-        Reply,                // agent -> client
-        Clipboard,            // both directions
-        DisplayConfig,        // client -> agent
-        AnnounceCapabilities, // both directions
-        ClipboardGrab,        // both directions
-        ClipboardRequest,     // both directions
-        ClipboardRelease,     // both directions
-        FileTransferStart,
-        FileTransferStatus,
-        FileTransferData,
-        Disconnected,
-        MaxClipboard,
-        VolumeSync,
-        GraphicsDeviceInfo,
-    };
 
-    enum class Capability {
-        MouseState = 0,
-        MonitorsConfig,
-        Reply,
-        Clipboard,
-        DisplayConfig,
-        ClipboardByDemand,
-        ClipboardSelection,
-        SparseMonitorsConfig,
-        GuestLineEndLF,
-        GuestLineEndCRLF,
-        MaxClipboard,
-        AudioVolumeSync,
-        MonitorsConfigPosition,
-        FileTransferDisabled,
-        FileTransferDetailedErrors,
-        GraphicsCardInfo,
-        ClipboardNoReleaseOnRegrab,
-        ClipboardGrabSerial,
-        __End,
+        // There are currently no messages which are meant for the server, so all messages sent by the agent (us) with this port are discarded.
+        Server
     };
 
-    enum class ClipboardType {
-        None = 0,
-        Text,
-        PNG,
-        BMP,
-        TIFF,
-        JPEG,
-        FileList,
-        __Count
-    };
+    static ErrorOr<NonnullOwnPtr<SpiceAgent>> create(StringView device_path);
+    SpiceAgent(NonnullOwnPtr<Core::File> spice_device, ConnectionToClipboardServer& clipboard_connection, Vector<Capability> const& capabilities);
 
-    constexpr static size_t CAPABILITIES_SIZE = ((size_t)Capability::__End + 31) / 32;
+    ErrorOr<void> start();
 
-    struct [[gnu::packed]] AnnounceCapabilities {
-        u32 request;
-        u32 caps[CAPABILITIES_SIZE];
+    template<typename T>
+    ErrorOr<void> send_message(T message)
+    {
+        // Attempt to write the message's data to a stream
+        auto message_stream = AK::AllocatingMemoryStream();
+        TRY(message.write_to_stream(message_stream));
 
-        static ByteBuffer make_buffer(bool request, Vector<Capability> const& capabilities);
-    };
+        // Create a header to be sent
+        auto header_stream = AK::AllocatingMemoryStream();
+        auto header = TRY(MessageHeader::create(message.type(), message_stream.used_buffer_size()));
+        TRY(header.write_to_stream(header_stream));
```

Comment:
Hm. Instead of creating two streams, could we use one stream where we write 0 as the message size at first, and then either manually or via `seek()` write the correct length value after the message is written?

Reply:
Yeah that makes a bit more sense, the two streams was definitely a bit confusing